### DISCUSSION TITLE
mainloop throwing error

### DISCUSSION THREAD

**Discussion by BhagyaJyoti22006** [TomSchimansky/CustomTkinter#2234] [Q&A] [answered]
I want to create a splash screen. in tkinter i can just call tkinter.mainloop() to run any added window. However ctk throws error:

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\py\code.py", line 77, in <module>
    ctk.mainloop()
    ^^^^^^^^^^^^
AttributeError: module 'customtkinter' has no attribute 'mainloop'

**Comment by Anonymous6598** [comment]
Show full code please

**Comment by BhagyaJyoti22006** [comment]
```
from customtkinter import *

splash = CTk()
splash.title("splash")
splash.geometry("300x200")

splash_label = CTkLabel(splash, text="splash screen text")
splash_label.place(x=4,y=8)

def window():
    splash.destroy()
    root=CTk()
    root.title("title")
    root.geometry("500x550")
splash.after(3000, window)

mainloop()
```

**Reply by Anonymous6598** [reply]
Just put `root.mainloop()` in the function.

**Comment by Anonymous6598** [comment]
You forgot to put 
```python
splash.mainloop()
```

**Comment by Anonymous6598** [comment]
That's why there was an error.

**Comment by Anonymous6598** [comment]
@BhagyaJyoti22006, did it work?

**Comment by BhagyaJyoti22006** [comment]
I am trying to have a splash screen work. therefore i put mainloop(). In tkinter, the code works with just mainloop().
I am creating a new window (splash_screen) and destroying it. Then, creating another window (main_window). Therefore, I used just mainloop() to govern any opened window. This worked with tkinter, thats why I am confused. 

splash.mainloop would only loop till splash is there. after its destroyed, no loop.

**Comment by Anonymous6598** [comment]
Try 
```python
root.splash()
```

**Comment by BhagyaJyoti22006** [comment]
well, root is not defined at that moment. 

Traceback (most recent call last):
  File "path", line 17, in <module>
    root.splash()
    ^^^^
NameError: name 'root' is not defined

If you meant root.mainloop(), that also throwing similar error.

**Comment by Anonymous6598** [comment]
Just put
 ```python
root.mainloop()
```
Inside the function.

**Comment by BhagyaJyoti22006** [comment]
root code wont work before splash destroys. So, without splash.mainloop, the code just closes without any window. 
Also, putting seperate mainloops for splash, root throws:

```
invalid command name "2816965442112update"
    while executing
"2816965442112update"
    ("after" script)
invalid command name "2816923050240check_dpi_scaling"
    while executing
"2816923050240check_dpi_scaling"
    ("after" script)
```

**Comment by Anonymous6598** [comment]
```python
from customtkinter import *
from sys import *

splash = CTk()
splash.title("splash")
splash.geometry("300x200")

splash_label = CTkLabel(splash, text="splash screen text")
splash_label.place(x=4,y=8)

def run():
    splash.mainloop()

def window():
    splash.destroy()
    root=CTk()
    root.title("title")
    root.geometry("500x550")
    root.protocol("WM_DELETE_WINDOW", exit)
    root.mainloop()
    
splash.after(3000, window)

run()
```
this might be solution to your problem

**Reply by Anonymous6598** [reply]
When you exit app, exit kills all processes, therefore mistake disappears.

**Comment by BhagyaJyoti22006** [comment]
Tried your code, still getting:

```
invalid command name "2048207648960update"
    while executing
"2048207648960update"
    ("after" script)
invalid command name "2048167844288check_dpi_scaling"
    while executing
"2048167844288check_dpi_scaling"
    ("after" script)
```

**Comment by Anonymous6598** [comment]
It disappears, when you exit program.

**Comment by Anonymous6598** [comment]
https://github.com/TomSchimansky/CustomTkinter/assets/121385046/f7194ce3-d8fe-4e71-aa31-221933bdd45a

**Comment by BhagyaJyoti22006** [comment]
Got it mate. But can we somehow get rid of that error message?

**Reply by Anonymous6598** [reply]
Unfortunately, not. Unless we use:
```python
async def window():
```

**Reply by Anonymous6598** [reply]
It might help

**Comment by Anonymous6598** [comment, marked as answer]
this is 100% solution. When you check it, mark it as an answer:
```python
from customtkinter import *
from sys import *

splash = CTk()
splash.title("splash")
splash.geometry("300x200")

splash_label = CTkLabel(splash, text="splash screen text")
splash_label.place(x=4,y=8)

def run_app():
    splash.mainloop()

def window():
    splash.withdraw()
    splash.quit()
    root=CTk()
    root.title("title")
    root.geometry("500x550")
    root.protocol("WM_DELETE_WINDOW", exit)
    root.mainloop()
    
splash.after(3000, window)

run_app()
#it disappears
```

**Reply by BhagyaJyoti22006** [reply]
Thanks a lot mate, the issue is disappeared now. How can I mark this as an answer? Apologies, never did it before.

**Reply by BhagyaJyoti22006** [reply]
Also, just for knowledge, could you please explain why this worked and what was throwing the earlier errors?

**Reply by Anonymous6598** [reply]
Now it quits first window and destroys window.
`quit()` - quits to another window
`withdraw()` - destroys first window

**Reply by Anonymous6598** [reply]
I also wanted to get an achievement from GitHub. But ok.

**Reply by BhagyaJyoti22006** [reply]
I cannot figure out how to mark your reply as an answer. Please guide?

**Reply by Anonymous6598** [reply]
There has to be button "mark as answer"

**Reply by Anonymous6598** [reply]
Below comment

**Reply by Anonymous6598** [reply]
Link to guide https://docs.github.com/en/discussions/collaborating-with-your-community-using-discussions/participating-in-a-discussion

**Reply by BhagyaJyoti22006** [reply]
I figured it out. The category was general, therefore the option was not coming. Marked it. Thanks again mate.

**Reply by Anonymous6598** [reply]
Thank you so much. I have received new achievement "galaxy brain". I'll become your follower.

**Reply by BhagyaJyoti22006** [reply]
congratulations!!! I'm happy for you.

**Reply by BhagyaJyoti22006** [reply]
> this is 100% solution. When you check it, mark it as an answer:
> 
> ```python
> from customtkinter import *
> from sys import *
> 
> splash = CTk()
> splash.title("splash")
> splash.geometry("300x200")
> 
> splash_label = CTkLabel(splash, text="splash screen text")
> splash_label.place(x=4,y=8)
> 
> def run_app():
>     splash.mainloop()
> 
> def window():
>     splash.withdraw()
>     splash.quit()
>     root=CTk()
>     root.title("title")
>     root.geometry("500x550")
>     root.protocol("WM_DELETE_WINDOW", exit)
>     root.mainloop()
>     
> splash.after(3000, window)
> 
> run_app()
> #it disappears
> ```


		
I am trying to do something different now. I have a window which has a button and a checkbox. Clicking on this button will close the window and if the checkbox is checked, the programme will reopen. But, similar error occured doing this.

while True:
	root = ctk.CTk()
	app = Programe(root)
	root.mainloop()
	if Info.refresh is False:
		break


The button calls self.root.destroy()


Error:

invalid command name "2784158436608update"
    while executing
"2784158436608update"
    ("after" script)
invalid command name "2784144862784check_dpi_scaling"
    while executing
"2784144862784check_dpi_scaling"
    ("after" script)
invalid command name "2784158566016_click_animation"
    while executing
"2784158566016_click_animation"
    ("after" script)

**Reply by Anonymous6598** [reply]
Got it.

**Reply by Anonymous6598** [reply]
Can you show full code?

**Reply by BhagyaJyoti22006** [reply]
> Can you show full code?
Yeah, I posted it below. Please check out.

**Comment by BhagyaJyoti22006** [comment]
```
import customtkinter as ctk

class Info:
    refresh = False

class App:
    def __init__(self, root):
        self.root = root
        self.root.geometry("200x300")
        self.root.protocol("WM_DELETE_WINDOW", self.close)
        self.button = ctk.CTkButton(self.root, command=self.refresh)
        self.button.grid(row=0,column=0)
    def refresh(self):
        Info.refresh = True
        self.root.destroy()
        pass
    def close(self):
        self.root.destroy()
        pass

def main():
    while True:
        root = ctk.CTk()
        app = App(root)
        root.mainloop()
        if Info.refresh is False:
            break
        # reload files and reopen dialog
        Info.refresh = False
    pass

if __name__ == "__main__":
    main()
```

This is my new code. If that button is clicked, the dialog is closed. Few files will be modified. And dialog is reopened.

**Reply by Anonymous6598** [reply]
You need exact same window reopen or not?

**Reply by BhagyaJyoti22006** [reply]
> You need exact same window reopen or not?

Yeah. I want that exact same window reopened.

**Comment by Anonymous6598** [comment]
this might be a solution:
```python
import customtkinter as ctk
import sys

class App(ctk.CTk):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        
        self.geometry("300x300")
        self.protocol("WM_DELETE_WINDOW", sys.exit)
        
        self.button = ctk.CTkButton(self, command=self.close)
        self.button.grid(row=0, column=0)
        
        self.checkbox = ctk.CTkCheckBox(self, text="reopen", onvalue="yes", offvalue="no")
        self.checkbox.grid(row=0, column=1)

    def close(self):
        self.checkboxvalueget = self.checkbox.get()
        if self.checkboxvalueget == "yes":
            self.withdraw()
            self.quit()
            
            self.newroot = ctk.CTk()
            self.newroot.geometry("300x300")
            self.newroot.protocol("WM_DELETE_WINDOW", sys.exit)
            self.newroot.mainloop()

        else:
            sys.exit()
        

def main():
    root = App()
    root.mainloop()

if __name__ == "__main__":
    main()
```

**Reply by BhagyaJyoti22006** [reply]
> this might be a solution:
> 
> ```python
> import customtkinter as ctk
> import sys
> 
> class App(ctk.CTk):
>     def __init__(self, *args, **kwargs):
>         super().__init__(*args, **kwargs)
>         
>         self.geometry("300x300")
>         self.protocol("WM_DELETE_WINDOW", sys.exit)
>         
>         self.button = ctk.CTkButton(self, command=self.close)
>         self.button.grid(row=0, column=0)
>         
>         self.checkbox = ctk.CTkCheckBox(self, text="reopen", onvalue="yes", offvalue="no")
>         self.checkbox.grid(row=0, column=1)
> 
>     def close(self):
>         self.checkboxvalueget = self.checkbox.get()
>         if self.checkboxvalueget == "yes":
>             self.withdraw()
>             self.quit()
>             
>             self.newroot = ctk.CTk()
>             self.newroot.geometry("300x300")
>             self.newroot.protocol("WM_DELETE_WINDOW", sys.exit)
>             self.newroot.mainloop()
> 
>         else:
>             sys.exit()
>         
> 
> def main():
>     root = App()
>     root.mainloop()
> 
> if __name__ == "__main__":
>     main()
> ```

I wanted the creation outside because I wanted asyncio codes. Is there any way to run asyncio-await codes inside tkinter? 
I should have made this clear beforehand. Apologies.

```
import customtkinter as ctk
import asyncio

class Info:
    refresh = False

class App:
    def __init__(self, root):
        self.root = root
        self.root.geometry("200x300")
        self.root.protocol("WM_DELETE_WINDOW", self.close)
        self.button = ctk.CTkButton(self.root, command=self.refresh)
        self.button.grid(row=0,column=0)
    def refresh(self):
        Info.refresh = True
        self.root.destroy()
        pass
    def close(self):
        self.root.destroy()
        pass

async def main():
    while True:
        root = ctk.CTk()
        app = App(root)
        root.mainloop()
        if Info.refresh is False:
            break
        Info.refresh = False
        # await function
    pass

if __name__ == "__main__":
    asyncio.run(main())
```

My logic is to close the window. run required await tasks and rerun window. Can this be achieved inside tkinter with your code? I tried few examples, but failed to run await codes inside tkinter.

**Reply by BhagyaJyoti22006** [reply]
After taking a look at your logic. I was able to acheive it as follows:-

```
import customtkinter as ctk
import asyncio

class Info:
    refresh = False

class App(ctk.CTk):
    def __init__(self):
        super().__init__()
        self.geometry("200x300")
        self.protocol("WM_DELETE_WINDOW", self.close)
        self.button = ctk.CTkButton(self, command=self.refresh)
        self.button.grid(row=0,column=0)
    def refresh(self):
        Info.refresh = True
        self.withdraw()
        self.quit()
        pass
    def close(self):
        self.withdraw()
        self.quit()
        pass

async def main():
    while True:
        root = App()
        root.mainloop()
        if Info.refresh is False:
            root.destroy()
            break
        Info.refresh = False
        # await function
    pass

if __name__ == "__main__":
    asyncio.run(main())
```

Thanks a lot once again bro!!!

**Reply by Anonymous6598** [reply]
You're welcome. Glad to see you happy.

**Reply by BhagyaJyoti22006** [reply]
> You're welcome. Glad to see you happy.

Sorry to trouble you but I got another error. I added an image to another button in that window. This gave an error which at the bottom wrote:
self.tk.call(_flatten((self._w, cmd)) + self._options(cnf))
_tkinter.TclError: image "pyimage2" doesn't exist

My code:-

```
import customtkinter as ctk
import asyncio
from PIL import Image

class Info:
    refresh = False

class App(ctk.CTk):
    def __init__(self):
        super().__init__()
        self.geometry("200x300")
        self.protocol("WM_DELETE_WINDOW", self.close)
        self.button = ctk.CTkButton(self, command=self.refresh)
        self.button.grid(row=0,column=0)
        self.image= Image.open("img/tag.png")
        self.image_ctk = ctk.CTkImage(light_image=self.image, dark_image=self.image, size=(32, 32))
        self.button2 = ctk.CTkButton(self, image=self.image_ctk)
        self.button2.grid(row=0,column=1)
        
    def refresh(self):
        Info.refresh = True
        self.withdraw()
        self.quit()
        pass
    def close(self):
        self.withdraw()
        self.quit()
        pass

async def main():
    while True:
        root = App()
        root.mainloop()
        if Info.refresh is False:
            root.destroy()
            break
        Info.refresh = False
        # await function
    pass

if __name__ == "__main__":
    asyncio.run(main())
```

You may use any example image for testing.

**Reply by Anonymous6598** [reply]
Code works. Just change name of photo, it might help

**Reply by Anonymous6598** [reply]
try this:
```python
import customtkinter as ctk
import asyncio
from PIL import Image

class Info:
    refresh = False

class App(ctk.CTk):
    def __init__(self):
        super().__init__()
        self.geometry("200x300")
        self.protocol("WM_DELETE_WINDOW", self.close)
        self.button = ctk.CTkButton(self, command=self.refresh)
        self.button.grid(row=0,column=0)
        self.image= Image.open("imagetag.png") # your different image file name
        self.image_ctk = ctk.CTkImage(light_image=self.image, dark_image=self.image, size=(32, 32))
        self.button2 = ctk.CTkButton(self, image=self.image_ctk)
        self.button2.grid(row=0,column=1)
        
    def refresh(self):
        Info.refresh = True
        self.withdraw()
        self.quit()
        pass
    def close(self):
        self.withdraw()
        self.quit()
        pass

async def main():
    while True:
        root = App()
        root.mainloop()
        if Info.refresh is False:
            root.destroy()
            break
        Info.refresh = False
        # await function
    pass

if __name__ == "__main__":
    asyncio.run(main())
```

**Reply by BhagyaJyoti22006** [reply]
> try this:

Same ' self.tk.call(_flatten((self._w, cmd)) + self._options(cnf))
_tkinter.TclError: image "pyimage2" doesn't exist ' error

I see that you only changed the image file name in this code. Maybe you gave me the wrong file.

**Reply by Anonymous6598** [reply]
There was an issue with file name. Name of the image was inappropriate. That's why you had KeyError.

**Reply by Anonymous6598** [reply]
Code is 100% correct.

**Reply by Anonymous6598** [reply]
Where is your image situated? Maybe it matters.

**Reply by Anonymous6598** [reply]
If you want, I can send you an image_example.py from customtkinter repo.

**Reply by Anonymous6598** [reply]
https://github.com/TomSchimansky/CustomTkinter/blob/master/examples/image_example.py

**Reply by BhagyaJyoti22006** [reply]
> Where is your image situated? Maybe it matters.

I renamed it to imagetag. Moved it to same directory as main.py. These are located in D drive. Still throwing rror:
self.tk.call(_flatten((self._w, cmd)) + self._options(cnf))
_tkinter.TclError: image "pyimage2" doesn't exist

Could you add any image and check if this error occurs for you as well?

**Reply by Anonymous6598** [reply]
I'll try.

**Reply by BhagyaJyoti22006** [reply]
> I'll try.

Thanks bro. I have now tried everything I could search. Now eagerly waiting for your next reply.

**Reply by Anonymous6598** [reply]
![Snimak ekrana 2024-08-18 181922](https://github.com/user-attachments/assets/dabf5025-09de-4f8e-8696-6e33ca231aa8)

**Reply by BhagyaJyoti22006** [reply]
> ![Snimak ekrana 2024-08-18 181922](https://private-user-images.githubusercontent.com/121385046/358921190-dabf5025-09de-4f8e-8696-6e33ca231aa8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwMDE2NzYsIm5iZiI6MTcyNDAwMTM3NiwicGF0aCI6Ii8xMjEzODUwNDYvMzU4OTIxMTkwLWRhYmY1MDI1LTA5ZGUtNGY4ZS04Njk2LTZlMzNjYTIzMWFhOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxOFQxNzE2MTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MjUyZjI5NDFjOGZjZmE2YjdkNWY0MmE1Y2JiNGZkMGRhM2EzMzY1ZGE1ODdlMTIyZTIzOWMxZDQxYWVmNDJkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.DOCW0pBvr38oqzRigB-dfNvmPJj-1oGw-rugIve9578)

Did you try clicking the first button and reopen app. the error occured during this procss only.

**Reply by BhagyaJyoti22006** [reply]
> I'll try.

I have created a new discussion thread: https://github.com/TomSchimansky/CustomTkinter/discussions/2543

Please join the discussion there and suggest what you think should be done. Others with similar problems would also be able to find their solution.
